### ISSUE TITLE
Orange LED blinks - but no buckets sniffed :(

### ISSUE THREAD

**Issue by hcgraf** [Portisch/RF-Bridge-EFM8BB1#76]
Hi!

I have a remote (Tevion GT-8000, using a HOLTEK HT48R01T3 chip) which is somehow received, but I'm not able to sniff anything…

- I enable bucket sniffing (rfraw 177), command is acknowledged (AAA055)
- pressing a button on the remote flashes the orange LED on the Sonoff
- But **no** buckets come back from the firmware :(

Other remotes work, so it's not a general problem with my setup.

Some web-search showed that the remote seems to be similar to Quigg GT-8000/GT-7000.
There might be a matching protocol description here: https://manual.pilight.org/protocols/433.92/switch/quigg_7000.html

Is that something that could be implemented here?
I'm happy to help if I can get some pointers :)

Thanks,
Hagen

**Comment by Portisch** [comment]
Try sniffing again with the latest [master](https://github.com/Portisch/RF-Bridge-EFM8BB1/blob/master/Keil%208051%20v9.53%20-%20Release/RF-Bridge-EFM8BB1.hex)!

I close the issue because of a buggy EFM8 firmware - please do a 0xB1 sniffing again and I will reopen the issue.

**Comment by Portisch** [comment]
The Limit is 0x7FFF because the highest bit is used as his high/low marking.
But you're right: https://github.com/Portisch/RF-Bridge-EFM8BB1/blob/ac0146a12fe853e6578ec7ba7e2c514c8a52975c/src/RF_Handling.c#L365
When the *10 get removed times up to 327670us should be possible.
I will check it!

**Comment by Portisch** [comment]
I took a look how this can be handled. The problem is when all the numbers geht smaller by factor 10 also the tolerances are getting smaller. Right now the firmware is trimmed for this tolerances.

When you have 10% tolerance and 380µs the tolerance will be +-38µs. When you have now 38(0)captures (1 captures  == 10µs) the tolerance is +-3(8)captures what is a difference of >20% to the tolerance before.
That's because I use integer calculation, not float. Float takes too much space and take too much time.

**Comment by hcgraf** [comment]
Thanks for looking into it!
First idea which comes to my mind: would it be acceptable (computationally) to use 32 bit integers instead of 16 bit?

If it's not too much effort:
Could you provide the factor 10 scaling (with reduced tolerances) as a branch?
Then I could look at the diff and maybe have another idea :)
(I'm not familiar with your codebase, so it's a bit difficult to reliably find all relevant places myself...)

Regards
Hagen

**Comment by Portisch** [comment]
I think I got it working. Please try with this hex the 0xB1 sniffing:
https://github.com/Portisch/RF-Bridge-EFM8BB1/blob/capture/Keil%208051%20v9.53%20-%20Release/RF-Bridge-EFM8BB1.hex

If it get now sniffed post the data, then a protocol can be implemented!

Tests:
`AA B1 03 013A 0380 2488 28181909081819081909090909081908190818181909090908 55`
Sync bucket: 2488 == 0x2488 * 10 == 93520µs

**Comment by hcgraf** [comment]
Great!
I hope I'll get some time later today to play around with it...
I'll let you know once I have something!
Thanks a lot already!

**Comment by hcgraf** [comment]
I tried to sniff some buttons: [scan.txt](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/2891103/scan.txt)

Remarks:
* It's quite "unreliable", i.e. I need to push a button approximately 5 times to get one result
* I tried to decode the buckets and send the data using 0xB1 command, but wasn't able to get a real receiver to react on it :(

**Comment by Portisch** [comment]
This is no useful data. You pressed the button a long time? There are at minimum 5 repeats needed and one transfer take a long time. So hold the button for minimum of 5s.

**Comment by hcgraf** [comment]
Oh sorry, I missed that information in your wiki…
Is this better: [scan.txt](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/2891268/scan.txt)?
It should contain 12 different buttons, one after the other pressed for some seconds.

**Comment by hcgraf** [comment]
Same 12 buttons, same order, but pressed even longer: 
[scan2.txt](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/2891289/scan2.txt)

this time I also kept some strange lines in there with "unknown commands". No idea where they come from…

**Comment by Portisch** [comment]
These scans are good enough. You have to take a look if some codes are repeated in the data. Then you have a good reading. Your sync time is 80270ys. I will check it tomorrow and give you some 0xB1 codes to test the transfer.

**Comment by Portisch** [comment]
Please make again some 0xB1 sniffes. For 2-4 buttons. But for each button a single scan file.
Now I don't know if there is a tolerance problem or just another button was pressed...

**Comment by hcgraf** [comment]
Ok, I made the full set of sniffs :) 
I pressed each button only once, but for >10s.
The remote has 4 "channels", with "On" and "Off" Buttons for each.
[Ch1Off.txt](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/2895460/Ch1Off.txt)
[Ch1On.txt](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/2895461/Ch1On.txt)
[Ch2Off.txt](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/2895462/Ch2Off.txt)
[Ch2On.txt](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/2895463/Ch2On.txt)
[Ch3Off.txt](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/2895464/Ch3Off.txt)
[Ch3On.txt](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/2895465/Ch3On.txt)
[Ch4Off.txt](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/2895466/Ch4Off.txt)
[Ch4On.txt](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/2895467/Ch4On.txt)

And it has one "On" and one "Off" for all channels simultaneously (some kind of wildcard address maybe).
[AllOff.txt](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/2895468/AllOff.txt)
[AllOn.txt](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/2895469/AllOn.txt)


Then it has something that might be strange:
A "Dimmer Plus" and "Dimmer Minus" button which are supposed to control the channel pressed that was pressed before (1-4, maybe it works even for "all" - can't check, my dimmers are broken :D)
I pressed CH4 before the Dimmer buttons.
[DimMinus.txt](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/2895473/DimMinus.txt)
[DimPlus.txt](https://github.com/Portisch/RF-Bridge-EFM8BB1/files/2895474/DimPlus.txt)

I'm going to try to make something out of the data now myself, but I'm always happy to get some hints :)

Thanks a lot!

**Comment by hcgraf** [comment]
Partial success :)
Strangely, the frequent lines in the scans **don't work** (after converting to 0xB0 codes & sending them).
However, the lines with "9" instead of "8" seem to work…
in AllOn.txt, line 5:
`20:28:28 MQT: tele/rf-bridge/RESULT = {"RfRaw":{"Data":"AA B1 04 007C 0057 0035 1F4F 380928090909090909090909090909090918181818 55"}}`
in Ch1On.txt, line 7:
`20:18:17 MQT: tele/rf-bridge/RESULT = {"RfRaw":{"Data":"AA B1 04 0079 0057 0035 1F66 381928090909090909090909092818180918181809 55"}}`

The bucket-lengths 0, 1, 2 are relatively similar (at least compared to bucket 3), 
maybe there's a bug assigning them correctly?
Or do you have any other explanation for this?

**Comment by Portisch** [comment]
This may because the different tolerance. Maybe another way to check if a bucket fit's in another needs another way to be defined.

Can you post a working 0xB0 code also please! May you are able to modify the bucket number from 4 down to 3 also?
Like 0x57 & 0x35 are almost the same.

So co

**Comment by hcgraf** [comment]
I edited the previous post & added the B0 data.
Will try combining buckets 0 & 1 now…

**Comment by Portisch** [comment]
from: `AA B0 1F 04 08 007C 0057 0035 1F4F 380928090909090909090909090909090918181818 55`
to: `AA B0 1D 03 08 007C 0057 1F4F 280918090909090909090909090909090918181818 55`
I removed bucket nr 2 because it is included only once!?

**Comment by hcgraf** [comment]
No, I think all 4 buckets are correct. Some are just assigned incorrectly.

By doing some pattern recognition on the **working** patterns, I found some pulses which are incorrectly placed into the ~0x57 bucket, but should be in the 0x7C (here actually 0x9B) bucket:
```
Original: AA B0 1F 04 08 0057 0035 009B 1F53 38081A080808080808080808080A0A0A0A0A0A0A0A 55
Changed : AA B0 1F 04 08 0057 0035 009B 1F53 3A081A282828282828282828282A0A0A0A0A0A0A0A 55
```
The original (sniffed & converted) doesn't work, but the changed one works!
I can't highlight in the code-view, but I changed the first '8' to 'A' and the middle '0's to '2'

**Comment by hcgraf** [comment]
> from: `AA B0 1F 04 08 007C 0057 0035 1F4F 380928090909090909090909090909090918181818 55`
> to: `AA B0 1D 03 08 007C 0057 1F4F 280918090909090909090909090909090918181818 55`
> I removed bucket nr 2 because it is included only once!?

Both of them work.

**Comment by Portisch** [comment]
Ok, than 0x35 is useless. I was wondering why this bucket got included only once.
Also strange is that this code is working:`AA B0 1D 03 08 007C 0057 1F4F 280918090909090909090909090909090918181818 55`
Because 2 is the long low sync. 80 is the first bit (high/low). But what is it? The bucket number of both is the same so 0x7c high and 0x7c low. Is it now 1 or 0??

**Comment by hcgraf** [comment]
I have a bunch of (fake-)arduinos somewhere :D, i.e. different kinds of ATmega, ATtiny are somewhere in a big box…
But no 433 MHz receiver for those, so that'll probably not help…

**Comment by hcgraf** [comment]
Hi,

I remembered the link I found in the very beginning:
https://manual.pilight.org/protocols/433.92/switch/quigg_7000.html
It seems to match the sniffing, only that the looooong pulse is in the end.

I tried to implement that protocol myself, but couldn't test it (I don't want to register with SiLabs just for that… is there a way to compile this project with open-source / registration-free tools?)
Could you please review this and maybe provide a binary? :)
https://github.com/hcgraf/RF-Bridge-EFM8BB1/tree/quigg_protocol

Thanks
Hagen

**Comment by Portisch** [comment]
I don't remember, is there a registration needed to download the setup?
I can't compile until Monday.

The commit is not correct!

You have to take your timing!
The start is the long bucket low(0x1F4F) and high(0x7c). The others are data bytes.
Look at the drawn graph by the bucket converter. You have every time low/high. First sync low/high than databytes low/high. You have to count it how much bits are included.

**Comment by hcgraf** [comment]
Is it important (for your firmware) if the long sync low is first or last?
For the radio protocol it's not...
Other than that, my sniffed numbers are close enough to 700us, 1400us and 81000us that I'd trust the link.

But I realized that 81000 doesn't fit into uint16_t :(
Can I fake it by using {LOW(2), LOW(2)} with bucket (2) = 40500us?
For sending that should work, but receiving might be tricky...

Regards
Hagen

**Comment by Portisch** [comment]
Take a look to my commit on this branch. You have to divide the timings by 10! So 140, 70 & 8100! This will fit in the uint16_t. ;)

**Comment by hcgraf** [comment]
Oh right... I blindly followed the wiki ^^
I'll update it later.

What about the single start pulse and the long silence as end instead of start?
Should that work, or should I rather make the silence part of the start sequence?

**Comment by Portisch** [comment]
Just make a repeat and you will see it - it doesn't make a difference.

**Comment by hcgraf** [comment]
Thanks for the comments! Added my replies there...
No time now to change the code, will do that later!

**Comment by hcgraf** [comment]
Hi again,

soo, I registered with SiLabs now and compiled the project myself:
https://github.com/hcgraf/RF-Bridge-EFM8BB1/tree/quigg_protocol

The TX part is working, the following are tested:
```
CH0 ON			AA A8 04 0E BF F1 10 55
CH0 OFF			AA A8 04 0E BF F0 00 55

ALL ON			AA A8 04 0E BF FF 00 55
ALL OFF			AA A8 04 0E BF FE 10 55
```
(can't test more, I have only one working receiver left ;) )

The bit-mapping matches https://manual.pilight.org/protocols/433.92/switch/quigg_7000.html, but I think it's actually **odd** parity. This mapping is also the reason I reverted the BIT0/BIT1 sequences, they are more logical like this (i.e. CH0 is 00, not 11).

Unfortunately, RX ("advanced sniffing", A6) doesn't seem to work.
Maybe related to the strange bucket-assignment we discovered already in my sniffing data?
Do you have any idea?

Regards
Hagen

**Comment by Portisch** [comment]
I added the protocol in the capture branch:
https://github.com/Portisch/RF-Bridge-EFM8BB1/commit/ec17634f9fac67e0e65e113dddb5fb43d937cda0
Please try if 0xA6 is now working. It is a little be different to yours.
Also check if the led is flashing. When the red led goes on it means a protocol got in sync.
If no data is getting received in tasmota the problem is the 0x35 bucket!? This will cancel the decoding.

**Comment by hcgraf** [comment]
No that also doesn't work. No LED activity at all.

But there might be something else wrong…
I'm not able to receive **anything** with A6 sniffing, not even with A4 sniffing.
That's also true for a remote that works with the standard protocol, i.e. after "rfraw 0", I get signals.
after "rfraw AA A4 55", nothing.

**Comment by hcgraf** [comment]
This protocol / remote is really crappy :(

I made A6 sniffing kind of work (https://github.com/hcgraf/RF-Bridge-EFM8BB1/commit/1cb8633da2cd3b01204a97e27a2083a0b23df27f)
by only checking that the sync bucket(s) duration is long enough (but can be longer) - this certainly makes sense for the LOW sync bucket, as this is just silence.

But still, the received data is not stable, and actually always slightly wrong (some bits are flipped).
So it seems that the tolerances of this protocol are really big…
Probably I'd need to get some other sniffing in place to really figure it out…

**Comment by Portisch** [comment]
Please try again a 0xB1 sniffing by the new firmware: https://github.com/Portisch/RF-Bridge-EFM8BB1/commit/109777dc7f83c0382d7a8b2e36daa0e129523cf5

I simulated your signal and I sniffed:
`AA B1 04 007C 0060 0036 1F45 381828090909090909090909090909090918181818 55`

And here the scope:
![scope_0](https://user-images.githubusercontent.com/3938570/53394811-a8bcb200-39a0-11e9-9da7-51c9133272b5.png)


As you can see with your sniffed data it is not possible to to decode the signal. High and low durations are equal. So I decreased the delta to find the buckets and hopefully now there will be more buckets sniffed from your remote.

**Comment by hcgraf** [comment]
No, with that decreased delta, it seems that a lot less is sniffed…
Nothing usable so far… I'll try something else now, stay tuned ;)

**Comment by hcgraf** [comment]
This seems to work quite nicely for A6 sniffing now:
https://github.com/hcgraf/RF-Bridge-EFM8BB1/commit/d5bc4319b553ce088268322c87c2659ca1b549bf

The idea is, that if the first sync pulse is low, comparing it to upper and lower bounds is not very smart.
There is upper bound of the pause between two transmissions, it's just silence.
I also reduced the lower bound quite a bit (only half of the protocol specification), that worked fine in my testing.
Maybe the 81ms in the Quigg protocol are too long, and quite often disturbed by noise?

**Comment by Portisch** [comment]
When I simulated it the sync wasn't the problem. The problem was the second bit because of out of tolerance. But I have to think about only checking one bucket of two. So you define only a high bucket for bit 0 and bit 1 and ignore the low ones...

**Comment by hcgraf** [comment]
No, I only ignore the low bucket **if it's the very first sync bucket.**
Because in that case, there could be silence before, so you don't know when the silence ends and the bucket starts ;)

**Comment by Portisch** [comment]
The sync are two buckets and two have to fit in tolerance and than the firmware starting decoding of the bits. I had no problem getting in sync. You will see it with a very short pulse of the red led. Because the second bit fails and turn of the led. May you have to open the case to see the flashing.

**Comment by hcgraf** [comment]
It's working fine now with my commit linked above ^^

I think the sniffing associated some buckets incorrectly, so it's probably difficult to reproduce the actual signal just from the sniffing data - if you don't have the actual remote…
